src/scene/materials/node-material.js

```diff
@@ -0,0 +1,692 @@
+        var connection = { srcIndex: srcIndex, srcVarName: srcVarName, dstIndex: dstIndex, dstVarName: dstVarName };
+
+        this.graphData.connections.push(connection);
+    },
+
+    _getGraphVarValueString: function (graphVar) {
+        var ret;
+
+        if (graphVar.type === 'float') {
+            ret = 'float(' + graphVar.valueX + ')';
+        } else if (graphVar.type === 'vec2') {
+            ret = 'vec2(' + graphVar.valueX + ', ' + graphVar.valueY + ')';
+        } else if (graphVar.type === 'vec3') {
+            ret = 'vec3(' + graphVar.valueX + ', ' + graphVar.valueY + ', ' + graphVar.valueZ + ')';
+        } else if (graphVar.type === 'vec4') {
+            ret = 'vec4(' + graphVar.valueX + ', ' + graphVar.valueY + ', ' + graphVar.valueZ + ', ' + graphVar.valueW + ')';
+        }
+
+        return ret;
+    },
+
+    _generateSubGraphCall: function (inNames, outNames) {
+        var i;
+        var graphVar;
+        var callString = '';
+
+        for (i = 0; i < this.graphData.graphVars.length; i++) {
+            graphVar = this.graphData.graphVars[i];
+            if (graphVar.name.startsWith('OUT_ret')) {
+                if (outNames[graphVar.name] != undefined) {
+                    callString += outNames[graphVar.name] + ' = ';
+                } else {
+                    // I guess this is actually valid (return value not assigned to anything)
+                }
+            }
+        }
+
+
+        if (this.graphData.customFuncGlsl) {
+            callString += this.name + '( ';
+        } else {
+            callString += this.name + '_' + this.id + '( ';
+        }
+
+        for (i = 0; i < this.graphData.graphVars.length; i++) {
+            graphVar = this.graphData.graphVars[i];
+            if (graphVar.name.startsWith('IN_')) {
+                if (inNames && inNames[graphVar.name] != undefined) {
+                    callString += inNames[graphVar.name] + ', ';
+                } else {
+                    // this is tricky - use varname and unique (temp) id
+                    callString += graphVar.name + '_' + this.id + ', ';
+                }
+            }
+        }
+
+        for (i = 0; i < this.graphData.graphVars.length; i++) {
+            graphVar = this.graphData.graphVars[i];
+            if (graphVar.name.startsWith('OUT_') && !graphVar.name.startsWith('OUT_ret')) {
+                if (outNames[graphVar.name] != undefined) {
+                    callString += outNames[graphVar.name] + ', ';
+                } else {
+                    // this shouldn't be possible - all outputs from connected subgraphs will have a tmpVar declared.
+                    // ERROR!!!!
+                }
+            }
+        }
+
+        if (callString.endsWith(', ')) callString = callString.slice(0, -2);
+
+        callString += ' );\n';
+
+        return callString;
+    },
+
+    _getGraphVarByName: function (name) {
+        // convienient but not fast - TODO: optimize?

```

Comment:
looks like var name is unique, so you can store them in a map of name->var. if you must maintain var order then keep the array you have and add a map of var name->index.

Reply:
Good idea - var name is indeed unique - so if we create a transient/temporary structure to accelerate this when needed - this would follow the pattern being used.

Reply:
I found that this function is not used currently - it is used by the shadergraph script interface - so I commented it our for now - and will revisit and optimize in the shadergraph script interface PR.